plugin/keyring_vault/vault_memory.h

```diff
@@ -0,0 +1,47 @@
+#ifndef MYSQL_VAULT_MEMORY
+#define MYSQL_VAULT_MEMORY
+
+#include <my_global.h>
+#include <string.h>
+#include "keyring_memory.h"
+
+namespace keyring
+{
+#if !defined(HAVE_STDCXX11) && !defined(WIN32)
+  inline void memset_s(void *dest, size_t dest_max, int c, size_t n)
+  {
+    volatile unsigned char *p = reinterpret_cast<unsigned char*>(dest);
+    while (dest_max-- && n--) {
+      *p++ = c;
+    }
+  } 
+#endif
+
+  template <class T> class Secure_allocator : public std::allocator<T>
```

Comment:
This is more like to not let a user of your `Secure_allocator`, to shoot himself in the  leg.  In other words, it's more about allowing something which should be forbidden.
Fir instance,
`std::string str = std::string("blah-blah", Secure_allocator());`
Here, `str` will still use default (non-secure) allocator despite the fact that you specified secure allocation explicitly via parameters.
The problem is that `Secure_allocator<T>` can be implicitly converted to 'std::allocator<T>' because of public inheritance.